pkg/router/controller/route_secret_manager_test.go

```diff
@@ -0,0 +1,1139 @@
+			},
+		},
+		&corev1.Secret{},
+		0,
+	)
+}
+
+func fakeSecret(namespace, name string, secretType corev1.SecretType, data map[string][]byte) *corev1.Secret {
+	return &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+		Data: data,
+		Type: secretType,
+	}
+}
+
+type fakePluginNew struct {
+	t      watch.EventType
+	route  *routev1.Route
+	err    error
+	stopCh chan struct{}
+}
+
+func (p *fakePluginNew) HandleRoute(t watch.EventType, route *routev1.Route) error {
+	defer close(p.stopCh)
+	p.t, p.route = t, route
+	return p.err
+}
+
+func (p *fakePluginNew) HandleNode(t watch.EventType, node *corev1.Node) error {
+	return fmt.Errorf("not expected")
+}
+
+func (p *fakePluginNew) HandleEndpoints(watch.EventType, *corev1.Endpoints) error {
+	return fmt.Errorf("not expected")
+}
+func (p *fakePluginNew) HandleNamespaces(namespaces sets.String) error {
+	return fmt.Errorf("not expected")
+}
+func (p *fakePluginNew) Commit() error {
+	return p.err
+}
+
+func TestRouteSecretManager(t *testing.T) {
+
+	scenarios := []struct {
+		name               string
+		route              *routev1.Route
+		secretManager      fake.SecretManager
+		eventType          watch.EventType
+		allow              bool
+		expectedRoute      *routev1.Route
+		expectedEventType  watch.EventType
+		expectedRejections map[string]string
+		expectedError      bool
+	}{
+		{
+			name: "route added without externalCertificate",
+			route: &routev1.Route{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "route-test",
+					Namespace: "sandbox",
+				},
+				Spec: routev1.RouteSpec{},
+			},
+			eventType: watch.Added,
+			expectedRoute: &routev1.Route{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "route-test",
+					Namespace: "sandbox",
+				},
+				Spec: routev1.RouteSpec{},
+			},
+			expectedEventType:  watch.Added,
+			expectedRejections: map[string]string{},
+			expectedError:      false,
+		},
+		{
+			name: "route added with externalCertificate denied",
+			route: &routev1.Route{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "route-test",
+					Namespace: "sandbox",
+				},
+				Spec: routev1.RouteSpec{
+					TLS: &routev1.TLSConfig{
+						ExternalCertificate: &routev1.LocalObjectReference{
+							Name: "tls-secret",
+						},
+					},
+				},
+			},
+			secretManager: fake.SecretManager{
+				Secret: fakeSecret("sandbox", "tls-secret", corev1.SecretTypeTLS, map[string][]byte{}),
+			},
+			eventType: watch.Added,
+			allow:     false,
+			expectedRoute: &routev1.Route{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "route-test",
+					Namespace: "sandbox",
+				},
+				Spec: routev1.RouteSpec{
+					TLS: &routev1.TLSConfig{
+						ExternalCertificate: &routev1.LocalObjectReference{
+							Name: "tls-secret",
+						},
+					},
+				},
+			},
+			expectedEventType: watch.Deleted,
+			expectedRejections: map[string]string{
+				"sandbox-route-test": "ExternalCertificateValidationFailed",
+			},
```

Comment:
You don't need to set all the expected values because you test case will fail before, at the error check.

Reply:
Test case will not fail, because we are expecting error from the test case by setting `expectedError: true`

So,

```go
			gotErr := rsm.HandleRoute(s.eventType, s.route)
			if (gotErr != nil) != s.expectedError {
				t.Fatalf("expected error to be %t, but got %t", s.expectedError, gotErr != nil)
			}
```
^^ the above `Fatalf` won’t be executed (even when  `expectedError: true`) and test will continue checking other expected values like `expectedRejections`. So, we need to define `expectedRejections` field as well.

Reply:
> test will continue checking other expected values like expectedRejections. So, we need to define expectedRejections field as well.

Why checking for expected values if the tested function is expected to fail?

Reply:
because we will have to make sure that the status of the route is getting updated correctly, when this plugin rejects any route.

https://github.com/openshift/router/blob/d23f540103830840b033ea33e8967289212ad52f/pkg/router/controller/route_secret_manager.go#L115-L120

^^ status is getting updated before returing with error